Improved test coverage

jamadden opened the pull request:
100% of everything except for 

- Things that will be removed (loggers) or probably removed/reworked (OverflowableBuffer)
- ftp/server.py
    
Aggregate coverage is 96%. There are a few small bug fixes mixed in as well as some minor project gardening.

There's a lot that's not covered in ftp/server.py yet, but I'll see what I can do.

mgedmin commented on src/zope/server/dualmodechannel.py:
```diff
@@ -208,4 +210,11 @@ def close(self):
+            # already closed (or never fully opened) it will
+            # raise a AttributeError because it tries to call close()
+            # on self.socket, which is None
+            pass
```
The only `self.socket = None` I see in my version of asyncore.dispatcher on Python 2.7 is in `__init__`!

jamadden replied:
Which is what I think is going on here: One of the tests is set up to throw an exception, and it does so from someplace before the object has been connected, so the socket is still None. And when it tries to close it--bam!

mgedmin commented on src/zope/server/ftp/publisher.py:
```diff
@@ -41,33 +41,24 @@ def type(self, path):
+    for _name in ('lsinfo', 'mtime', 'size', 'mkdir', 'remove', 'rmdir'):
+        f = locals()[_name] = lambda self, path, _name=_name: self._execute(path, _name)
+        f.__name__ = _name
+    del _name
```
I find the old repetetive code much more readable.

jamadden replied:
So the issue with the old code was that the methods were all identical, except for one embedded string constant. Worse, only half of them were actually called. I wasn't up to writing meaningful tests for the half that weren't called, and I really didn't like the repetition.

One other option would be to do this dynamically in `__getattr__`, but that would break introspection of the class. Another option would be to do it after the class body, which is more verbose, but eliminates the use of `locals`, which I know some people find objectionable (even though this same pattern is found in the stdlib).

mgedmin commented on src/zope/server/ftp/server.py:
```diff
@@ -837,41 +838,22 @@ def reportDefault(self):
+        self.channel.finishSTOR(self.inbuf, self.finish_args)
 
+    def finish(self):
+        "Does nothing"
```
Tripple `"""` for docstrings please?

jamadden replied:
Sure, will do (everywhere you caught me!). Some of these are mine and some are existing code that was moved around. This package isn't terrible consistent when it comes to that.

mgedmin commented on src/zope/server/ftp/tests/test_ftpserver.py:
```diff
@@ -121,10 +121,10 @@ def loop(self):
                 print("EXCEPTION POLLING IN LOOP(): %s" % data)
-                if data[0] == EBADF:
-                    for key in asyncore.socket_map.keys():
+                if data.args[0] == EBADF:
```
`data.errno` ought to be more readable.

mgedmin commented on src/zope/server/http/wsgihttpserver.py:
```diff
@@ -136,19 +137,24 @@ def executeRequest(self, task):
+    def post_mortem(cls, exc_info):
+        import pdb
+        print("%s:" % exc_info[0])
+        print(sys.exc_info()[1])
```
I think you want `exc_info`, not `sys.exc_info()` here.

mgedmin replied:
Personally I would also prefer something like

```python
print("{0.__class__.__name__}: {1}".format(*exc_info))
```

(in terms of output formatting, not necessarily this exact spelling.)

jamadden replied:
Good point about the variable vs method call. That was missed during the refactoring.

In practice, this is called from an exception handler so it doesn't make a difference right now, but it makes sense to fix it.

mgedmin commented on src/zope/server/linereceiver/tests/test_lineserverchannel.py:
```diff
@@ -0,0 +1,110 @@
+        class Chunnel(Channel):
+            authenticated = True
+            # Python 3.4 includes a __getattr__ in asyncore.dispatcher
+            # that does `gettar(self.socket, name)`, but self.socket is only
```
`gettar` seems to be a misspelling of `getattr`.

Note that Python 3.6 no longer defines a `__getattr__` in asyncore.py.  I had to make one change because of that somewhere.

jamadden replied:
Right, that went away in 3.5. It's not documented, it just happened. The bug report claims it had been deprecated for some time (and there's a change in trigger.py that came over from ZEO dealing with the deprecation.)

Good catch on the spelling.

mgedmin commented on src/zope/server/maxsockets.py:
```diff
@@ -11,80 +11,24 @@
+
+.. caution:: These functions are all hard coded to return a constant value
+   of 100. Attempting to dynamically compute them by looping and creating
+   sockets until failure was very time consuming.
```
The original comment gave me the impression it was very time consuming *on MacOS X*.  Maybe let's preserve that bit of information?

jamadden replied:
gevent had or had something similar and I observed it to be slow (and buggy) everywhere, or at least both locally on OS X and on Travis containers.

mgedmin commented on src/zope/server/serverchannelbase.py:
```diff
@@ -103,9 +105,11 @@ def kill_zombies(self):
-        for channel in self.active_channels.values():
+        # channel.close calls channel.del_channel, which can change
+        # the size of the map.
+        for channel in list(self.active_channels.values()):
```
Ooh, subtle!  Good catch!

jamadden replied:
It's less subtle when it blows up in your face 😄 One of the new tests triggered that.

mgedmin commented on src/zope/server/task.py:
```diff
@@ -0,0 +1,78 @@
+        raise NotImplementedError()
+
+    def cancel(self):
+        'See ITask'
```
`"""` for docstrings?

mgedmin commented on src/zope/server/task.py:
```diff
@@ -0,0 +1,78 @@
+        self.channel.close_when_done()
+
+    def defer(self):
+        'See ITask'
```
`"""` for docstrings?

mgedmin commented on src/zope/server/tests/asyncerror.py:
```diff
@@ -35,7 +35,7 @@ def setUp(self):
 
     def tearDown(self):
-        self.unhook_asycnore_error()
+        self.unhook_asyncore_error()
```
Had to stare at this a long time until I noticed the difference, haha!

mgedmin commented on src/zope/server/trigger.py:
```diff
@@ -123,106 +120,126 @@ def handle_read(self):
+            # self.socket.fd. This breals the instantiation semantics and
+            # is a bug imo.  I dount it will get fixed, but maybe
+            # it will. Who knows. For that reason, we test for the
+            # fd changing rather than just checking the Python version.
```
Typos: breals -> breaks; dount -> doubt.

jamadden replied:
Good catch, easily fixed.

This was [copied from ZEO](https://github.com/zopefoundation/ZEO/commit/f4cf6a533f7be5c57d34d868440b18ab37353696), which no longer uses triggers. (But it had accumulated some fixes that had yet to be copied back to here.)

mgedmin commented on src/zope/server/trigger.py:
```diff
@@ -123,106 +120,126 @@ def handle_read(self):
+                self._connect_client(w, connect_address)
+                break    # success
+            except socket.error as detail:
+                if detail.args[0] not in self.ADDR_IN_USE_CODES:
```
`detail.errno`?

jamadden replied:
The thing about that is that it's not possible to set errno in the constructor on 2.7 or 3.x; you have to manually set it:
```pycon
In [6]: import socket

In [7]: socket.error
Out[7]: OSError

In [8]: socket.error(1).errno

In [9]: socket.error(1).args
Out[9]: (1,)
```
This feels safer to me, I guess?

mgedmin replied:
> it's not possible to set errno in the constructor on 2.7 or 3.x

This turns out not to be the case!  You just have to provide a string message _in addition_ to the errno:

```
>>> socket.error(42, "a message").errno
42
```

I tested this on Python 2.7, 3.4, 3.5, 3.6.

mgedmin commented on src/zope/server/trigger.py:
```diff
@@ -123,106 +120,126 @@ def handle_read(self):
+                if detail.args[0] not in self.ADDR_IN_USE_CODES:
+                    # "Address already in use" is the only error
+                    # I've seen on two WinXP Pro SP2 boxes, under
+                    # Pythons 2.3.5 and 2.4.1.
```
Idea: git blame to discover who is the "I" in this comment and add an attribution.

(This assumes the original attribution hasn't already been lost.)

jamadden replied:
It appears to be Tim Peters https://github.com/zopefoundation/ZEO/commit/c4f736a78ca6713fc3dec21f8aa1fa6f144dd82f

jamadden commented:
Thanks for wading through this. It turned out to be more work than I expected, so I know it's a lot to review. 

I got the FTP server to 100% without too many `no cover` directives (I hope)---and those that do exist are almost entirely in error handling cases I couldn't easily figure out how to trigger.

jamadden commented:
I'm going to go ahead and merge this since it was approved, I addressed most all the feedback, and I think the only thing left being discussed is a style point. That way I can cleanly move on to the other outstanding issues. If you'd like to continue discussing the remaining open point or request changes, I'll be happy to do so!

mgedmin commented:
Thank you for taking this on!
